arch: common: semihosting: file operations

JordanYates opened the pull request:
Implements the ARM semihosting mechanism for performing file operations on the host system when a device is emulated or running under a debugger. The RISC-V architecture also implements semihosting using ARM's API.

One use case for this feature is when attempting to validate code on datasets larger than those that can fit in the ROM of an emulated image. With this feature, arbitrarily large files can simply be read from the host filesystem.

The file operations are not implemented as standard library functions (_open, _lseek, etc) as the arguments and behaviours of the ARM mechanism do not map cleanly to the standard posix functions. 

For example the `open` call returns a file handle (like posix open) but takes a non-standard integer for the file mode, which corresponds to strings passed to the posix fopen call.

This work also extends `CONFIG_CONSOLE_SEMIHOST` to work on all emulated ARM architectures, not just Cortex-M3.

More information on semihosting can be found here:
https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-
Register and exception call semantics can be found here:
https://github.com/ARM-software/abi-aa/blob/main/semihosting/semihosting.rst

Commit:
arch: common: dedicated `SEMIHOST` symbol

Control the usage of semihosting with a dedicated symbol, instead of
implying semihosting from the usage of `SEMIHOST_CONSOLE`. This allows
semihosting to be used without the semihost console.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
arch: common: semihost: add semihosting operations

Add an API that utilizes the ARM semihosting mechanism to interact with
the host system when a device is being emulated or run under a debugger.

RISCV is implemented in terms of the ARM implementation, and therefore
the ARM definitions cross enough architectures to be defined 'common'.

Functionality is exposed as a separate API instead of syscall
implementations (`_lseek`, `_open`, etc) due to various quirks with
the ARM mechanisms that means function arguments are not standard.

For more information see:
https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

impl

Commit:
console: semihost_console: use semihost API

Update the semihost_console implementation to use the semihost API
instead of manually constructing the supervisor calls.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: lib: semihost: test file operations

Add tests for basic file operations under semihosting.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

nashif commented:
shouldn't this go into its own subsys instead of a generic library/os component that is usually reserved for cross-arch features?

gmarull commented on include/sys/semihost.h:
```diff
@@ -0,0 +1,121 @@
+ * @param fd handle returned by @ref semihost_open.
+ * @param offset offset from the start of the file in bytes.
+ *
+ * @retval 0 on success.
+ * @retval negative error code on failure.
```
nit: format is `@retval VALUE DESCRIPTION`, so maybe use `@retval -errno Negative error...`

de-nordic replied:
Would that be host `errno` codes or Zephyr `errno` codes?

JordanYates commented:
> shouldn't this go into its own subsys instead of a generic library/os component that is usually reserved for cross-arch features?

Ideally the function implementations belong in a file common to `ARM` and `ARM64`, with the supervisor call done in arch specific code. Making this a generic subsys doesn't make much sense IMO because any other architectures are probably going to have their own bespoke API's of what they support.

Commit:
tests: console: extend semihost_console testing

With SEMIHOST_CONSOLE now being supported on all ARM, ARM64 and RISC-V
architectures, extend the testing to cover these cases.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
boards: arm: qemu_cortex_a9: move UART_CONSOLE

Move where the default value of `CONFIG_UART_CONSOLE` is set from
`Kconfig.defconfig` to `qemu_cortex_a9_defconfig`. This conforms to
the standard location and lets the default be overridden by
applications.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Implementation has been moved to be under `arch`.
Semihosting is now supported for all ARM targets, not just Cortex-M3

nordic-krch commented on cmake/emu/qemu.cmake:
```diff
@@ -70,7 +70,7 @@ endif()
 
 # Connect semihosting console to the console chardev if configured.
-if(CONFIG_SEMIHOST_CONSOLE)
+if(CONFIG_ARM_SEMIHOST)
```
is this change needed?

JordanYates replied:
Yes, `-semihosting-config enable=on` is needed for semihosting to work

nordic-krch commented on include/arch/common/arm_semihost.h:
```diff
@@ -0,0 +1,190 @@
+ * @param instr instruction code to run
+ * @param args instruction specific arguments
+ *
+ * @return integer return code of instruction
```
any more details what might be ok value and which is an error? Many functions bellow return directly after calling this function.

JordanYates replied:
Not really, this function is only exposed in case a user wants to run one of the other instructions. The meanings of the return codes varies by instruction, and would need to be  checked per instruction here: https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-

microbuilder commented:
Can you add a sample application (doesn't need to be complex) demonstrating how this can be used, along with a short README point people to the API documentation, etc.

Ran the tests on a Cortex-M device (`$ ./scripts/twister -p mps2_an521 -N --inline-logs -T tests/arch/arm/arm_semihost`) and they seem to pass, but a sample and a bit of documentation would help people understand how to use this more effectively, and bring it to their attention.

JordanYates commented:
> Can you add a sample application (doesn't need to be complex) demonstrating how this can be used, along with a short README pointing people to the API documentation, etc.

I can, but the sample cannot really do anything more than what the sample is performing (we have no knowledge of the install path of Zephyr or the build directory). So all that can really be done is make a file, write some data to it, and read it back out (what the test does).

microbuilder commented:
> I can, but the sample cannot really do anything more than what the sample is performing (we have no knowledge of the install path of Zephyr or the build directory). So all that can really be done is make a file, write some data to it, and read it back out (what the test does).

Perhaps just add some documentation to describe the API, then, and the links in this PR so that it gets into the rendered documentation.

Commit:
doc: guides: semihost: added

Add a guide section on how to use semihosting with an example code
section on opening a file to read data from it.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
> Perhaps just add some documentation to describe the API, then, and the links in this PR so that it gets into the rendered documentation.

Done

JordanYates commented:
With the merge of #44566 this has been extended to cover semihosting for ARM, ARM64 and RISC-V platforms.
Conveniently RISC-V semihosting is explicitly following the ARM API.
IMO this is enough to make the ARM implementation a "common" one.

stephanosio commented on drivers/console/Kconfig:
```diff
@@ -298,17 +298,12 @@ config NATIVE_POSIX_CONSOLE_INIT_PRIORITY
 	bool "Use semihosting for console"
 	select CONSOLE_HAS_DRIVER
-	depends on CPU_CORTEX_M || ARM64 || RISCV
+	depends on ARM_SEMIHOST || RISCV
```
`ARM_SEMIHOST` is not a valid Kconfig symbol. Did you mean `SEMIHOST`?

stephanosio commented on tests/drivers/console/prj_semihost.conf:
```diff
@@ -2,4 +2,5 @@ CONFIG_TEST=y
 CONFIG_CONSOLE=y
 CONFIG_SERIAL=n
 CONFIG_UART_CONSOLE=n
+CONFIG_ARM_SEMIHOST=y
```
same here

stephanosio commented on include/zephyr/arch/common/semihost.h:
```diff
@@ -0,0 +1,200 @@
+ * @retval 0 on success.
+ * @retval -errno negative error code on failure.
+ */
+long semihost_write(long fd, void *buf, long len);
```
`const void *buf`

stephanosio commented on arch/common/semihost.c:
```diff
@@ -0,0 +1,131 @@
+#include <zephyr/kernel.h>
+#include <zephyr/arch/common/semihost.h>
+
+struct _semihost_poll_in_args {
```
Please do not add any new identifiers that start with `_` (i.e. reserved identifiers).

stephanosio reviewed (changes requested):
Looks ok overall.

Just some minor comments.

stephanosio reviewed (approved):
LGTM

bbolen commented on tests/arch/common/semihost/src/main.c:
```diff
@@ -0,0 +1,76 @@
+static void test_file_ops(void)
+{
+	const char *test_file = "./test.bin";
+	uint8_t w_buffer[16] = { 1, 2, 3, 4, 5 };
```
this could be made const
